onmt/utils/loss.py

```diff
@@ -172,21 +175,136 @@ def sharded_compute_loss(self, batch, output, attns,
             batch_stats.update(stats)
         return batch_stats
 
-    def _stats(self, loss, scores, target):
+    def _stats(self, loss, scores, target, batch_sz=None):
         """
         Args:
             loss (:obj:`FloatTensor`): the loss computed by the loss criterion.
             scores (:obj:`FloatTensor`): a score for each possible output
             target (:obj:`FloatTensor`): true targets
+            batch_sz (:'int'): batch size
 
         Returns:
             :obj:`onmt.utils.Statistics` : statistics for this batch.
         """
         pred = scores.max(1)[1]
+        precision_matches = Counter()
+        precision_totals = Counter()
+        prediction_lengths = 0
+        reference_lengths = 0
+
+        if batch_sz is not None:
```

Comment:
don't really understand why we need the batch_size for.

Reply:
_get_modified_precision_counts assumes the shape as (batch,seq_length) before this it is not that.

Reply:
what I meant is that I don't really see why you need to do the calculation here in the loss.py module.
it should only happen in the statistics.py
as I said before, you need to keep all the Bleu classes / module in its own bleupy so that it can be reused elsewhere, but so far we just need Bleu as a stat, not a trigger for training.